substrate/network-libp2p/src/service.rs

```diff
@@ -0,0 +1,1106 @@
+						upgrade::apply(out.socket, kad_upgrade.clone(),
+							endpoint, client_addr)
+					)
+					.map(move |(kad_ctrl, stream), _| {
+						if let Some(tx) = tx.lock().take() {
+							let _ = tx.send(kad_ctrl.clone());
+						}
+						(FinalUpgrade::Kad((kad_ctrl, stream)), Endpoint::Dialer)
+					})
+				)
+				.expect("cannot dial");
+			rx.map_err(|err| IoError::new(IoErrorKind::ConnectionAborted, err))
+		})
+		.into_future()
+		.and_then(|(_peer_id, kad_ctrl)| kad_ctrl);
+
+	// Note that we use a Box in order to speed compilation time.
+	Box::new(final_future) as Box<Future<Item = _, Error = _>>
+}
+
+/// Processes the information about a node.
+///
+/// - `original_addr` is the address used to originally dial this node.
+/// - `endpoint` is whether we dialed or listened to this node.
+/// - `transport` is used for the `nat_traversal` method.
+///
+/// Returns an error if the node has been refused access.
+fn process_identify_info(
+	shared: Arc<Shared>,
+	info: &IdentifyTransportOutcome,
+	original_addr: Multiaddr,
+	endpoint: Endpoint,
+	transport: &impl Transport
+) -> Result<(), IoError> {
+	let node_id = info.info.public_key.clone().into_peer_id();
+	let _peer_id = shared.network_state.set_peer_info(node_id.clone(), endpoint,
+		info.info.agent_version.clone(), original_addr.clone(),
+		original_addr.clone())?;	// TODO: wrong local addr
+
+	if let Some(ref original_listened_addr) = *shared.original_listened_addr.read() {
+		if let Some(ext_addr) = transport.nat_traversal(original_listened_addr, &info.observed_addr) {
+			let mut listened_addrs = shared.listened_addrs.write();
+			if !listened_addrs.iter().any(|a| a == &ext_addr) {
+				trace!(target: "sub-libp2p", "NAT traversal: remote observes us as \
+					{} ; registering {} as one of our own addresses",
+					info.observed_addr, ext_addr);
+				listened_addrs.push(ext_addr);
+			}
+		}
+	}
+
+	for addr in info.info.listen_addrs.iter() {
+		trace!(target: "sub-libp2p", "Peer store: adding address {} for {:?}",
+			addr, node_id);
+		shared.network_state.add_kad_discovered_addr(&node_id, addr.clone());
+	}
+
+	Ok(())
+}
+
+/// Expects a multiaddr of the format `/p2p/<node_id>` and returns the node ID.
+/// Panics if the format is not correct.
+fn p2p_multiaddr_to_node_id(client_addr: Multiaddr) -> PeerstorePeerId {
+	let (first, second);
+	{
+		let mut iter = client_addr.iter();
+		first = iter.next();
+		second = iter.next();
+	}
+	match (first, second) {
+		(Some(AddrComponent::P2P(node_id)), None) =>
+			PeerstorePeerId::from_bytes(node_id).expect("libp2p always reports a valid node id"),
+		_ => panic!("Reported multiaddress is in the wrong format ; programmer error")
+	}
+}
+
+/// Since new protocols are added after the networking starts, we have to load the protocols list
+/// in a lazy way. This is what this wrapper does.
+#[derive(Clone)]
+struct DelayedProtosList(Arc<Shared>);
+impl<C, Maf> ConnectionUpgrade<C, Maf> for DelayedProtosList
```

Comment:
what is `Maf` short for?

Reply:
`MultiaddressFuture`. Should I add a comment?

Reply:
yes, that would be helpful